specification/at_protocol_specification.md

```diff
@@ -0,0 +1,1110 @@
+
+<!-- ## Life Cycle -->
+
+<!-- <TO DO: TBD: What I want to write is, A User of the secondary should be able to control pausing and resuming of a Secondary Server> -->
+
+## Verbs
+
+### The `from` verb
+
+**Synopsis:**
+
+The `from` verb is used to tell a secondary whom you claim to be.
+
+Following regex represents the syntax of the `from` verb:
+
+**Example:**
+
+```r'^from:(?<@sign>@?[^@\s]+$)' ```
+
+**Response:**
+
+If the user who is trying to connect is the owner of the Secondary Server, then the `from` verb should respond with the following response.
+
+```key:<sessionId@sign:uuid>```
+
+If the user who is trying to connect is not the owner of the Secondary Server, then the `from` verb should respond with the following response.
+
+```proof:<sessionid>@<@sign>:<UUID>```
+
+If the user is not allowed to connect to the Secondary Server, then it should respond back with the following error and close the connection to the server.
+
+```error:AT0013-Connection Exception```
+
+**Description:**
+
+The `from` verb is used to tell the Secondary Server what @sign you claim to be. With the `from` verb, one can connect to one's own Secondary Server or someone else's Secondary Server. In both cases, the Secondary Server responds back with a challenge to prove that you are who you claim to be. This is part of the authentication mechanism of the @protocol. 
+
+This authentication mechanism varies based on whether you are connecting to your own secondary (cram) or someone else's secondary (pol).
+
+**OPTIONS:**
+
+```<@sign>``` 
+Required: Yes
+Description: @sign with which you are connecting to a Secondary Server.
+
+### The `cram` verb
+
+**Synopsis:**
+
+The `cram` verb is used to boostrap authenticate one's own self as an owner of a Secondary Server. It is 
+intended to be used once until a set of PKI keys are cut on the owner's mobile device and from then on 
+we use the `pkam` verb.
+
+The following regex represents the syntax of the `cram` verb:
+
+```r'^cram:(?<digest>.+$)'```
+
+**Response:**
+
+If the user gets the challenge right, the prompt should change to the @sign of the user.
+
+```<@sign>@```
+
+If the user gets the cram authentication wrong, then it should respond back with the following error and close the connection to the server.
+
+```error:AT0401-Client authentication failed```
+
+**Description:**
+
+The `cram` verb follows the `from` verb. As an owner of the Secondary Server, you should be able to take the challenge thrown by the `from` verb and encrypt using the shared key that the server has been bound with. Upon receiving the `cram` verb along with the digest, the server decrypts the digest using the shared key and matches it with the challenge. If they are the same, then the secondary lets you connect to the Secondary Server and changes the prompt to your @sign.
+
+**OPTIONS:**
+
+```<digest> ```
```

Comment:
Is this meant to be on one line? It doesn't scan very well at the moment. Maybe use a table?